{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Let's import a bunch of stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import seaborn as sb\n",
    "import scanpy.api as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.sparse import csr_matrix\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "from sklearn import preprocessing\n",
    "from sklearn.preprocessing import MaxAbsScaler\n",
    "import argparse\n",
    "import logging\n",
    "import gc\n",
    "from skbio.stats.composition import clr\n",
    "%matplotlib inline\n",
    "# Savefile path\n",
    "savepath = 'V6.7.flareAb.h5ad'\n",
    "saveno_norm_path = 'FlareDataset.h5ad'\n",
    "pd.set_option('display.max_rows', 300)\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.width', 500)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import data file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RUN_WELL</th>\n",
       "      <th>ID</th>\n",
       "      <th>ALIAS</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>TR/UNT</th>\n",
       "      <th>PAIR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_1</td>\n",
       "      <td>0_904194200_904194200</td>\n",
       "      <td>FLARE2,3,or8</td>\n",
       "      <td>UCSF</td>\n",
       "      <td>FLARE</td>\n",
       "      <td>Unknown (2 TR, 1 UNT)</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_1</td>\n",
       "      <td>0_1716_1716</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_1</td>\n",
       "      <td>0_IGTB256</td>\n",
       "      <td>IGTB256</td>\n",
       "      <td>IMMVAR</td>\n",
       "      <td>HEALTHY</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_1</td>\n",
       "      <td>0_1892_1892</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_1</td>\n",
       "      <td>0_900805200_900805200</td>\n",
       "      <td>1804_1804</td>\n",
       "      <td>CLUES</td>\n",
       "      <td>HEALTHY</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1_1</td>\n",
       "      <td>0_904236200_904236200</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1_1</td>\n",
       "      <td>0_Criswell_Ye_E04_1941DNA01_Criswell_Ye_E04_19...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1_1</td>\n",
       "      <td>0_1754_1754</td>\n",
       "      <td>1754_1754</td>\n",
       "      <td>CLUES</td>\n",
       "      <td>HEALTHY</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1_1</td>\n",
       "      <td>0_IGTB1921</td>\n",
       "      <td>IGTB1921</td>\n",
       "      <td>IMMVAR</td>\n",
       "      <td>HEALTHY</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1_1</td>\n",
       "      <td>0_1763_1763</td>\n",
       "      <td>FLARE2,3,or8</td>\n",
       "      <td>UCSF</td>\n",
       "      <td>FLARE</td>\n",
       "      <td>Unknown (2 TR, 1 UNT)</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  RUN_WELL                                                 ID         ALIAS  SOURCE   STATUS                 TR/UNT  PAIR\n",
       "0      1_1                              0_904194200_904194200  FLARE2,3,or8    UCSF    FLARE  Unknown (2 TR, 1 UNT)  True\n",
       "1      1_1                                        0_1716_1716             -       -        -                      -     -\n",
       "2      1_1                                          0_IGTB256       IGTB256  IMMVAR  HEALTHY                      -     -\n",
       "3      1_1                                        0_1892_1892             -       -        -                      -     -\n",
       "4      1_1                              0_900805200_900805200     1804_1804   CLUES  HEALTHY                      -     -\n",
       "5      1_1                              0_904236200_904236200             -       -        -                      -     -\n",
       "6      1_1  0_Criswell_Ye_E04_1941DNA01_Criswell_Ye_E04_19...             -       -        -                      -     -\n",
       "7      1_1                                        0_1754_1754     1754_1754   CLUES  HEALTHY                      -     -\n",
       "8      1_1                                         0_IGTB1921      IGTB1921  IMMVAR  HEALTHY                      -     -\n",
       "9      1_1                                        0_1763_1763  FLARE2,3,or8    UCSF    FLARE  Unknown (2 TR, 1 UNT)  True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_excel('flare.covars.again.xlsx')\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Combine flare samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "Initial\n",
      "AnnData object with n_obs × n_vars = 12401 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "SNG confirmation via SNG- tag in BEST\n",
      "View of AnnData object with n_obs × n_vars = 8948 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Removing barcodes with no covariate data\n",
      "View of AnnData object with n_obs × n_vars = 8948 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:61: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:62: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:66: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:64: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Structure details: AnnData object with n_obs × n_vars = 8948 × 32758 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'ind_cov'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "Initial\n",
      "AnnData object with n_obs × n_vars = 12648 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "SNG confirmation via SNG- tag in BEST\n",
      "View of AnnData object with n_obs × n_vars = 9458 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Removing barcodes with no covariate data\n",
      "View of AnnData object with n_obs × n_vars = 9458 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:115: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:116: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:117: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:121: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Structure details: AnnData object with n_obs × n_vars = 18406 × 32758 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'batch', 'batch_cov', 'ind_cov', 'well'\n",
      "    var: 'gene_ids-0', 'protein?-0', 'gene_ids-1', 'protein?-1'\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "Initial\n",
      "AnnData object with n_obs × n_vars = 12635 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "SNG confirmation via SNG- tag in BEST\n",
      "View of AnnData object with n_obs × n_vars = 7789 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Removing barcodes with no covariate data\n",
      "View of AnnData object with n_obs × n_vars = 7789 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Structure details: AnnData object with n_obs × n_vars = 26195 × 32758 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'batch', 'batch_cov', 'ind_cov', 'well'\n",
      "    var: 'gene_ids-0-0', 'protein?-0-0', 'gene_ids-1-0', 'protein?-1-0', 'gene_ids-1', 'protein?-1'\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "Initial\n",
      "AnnData object with n_obs × n_vars = 12877 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "SNG confirmation via SNG- tag in BEST\n",
      "View of AnnData object with n_obs × n_vars = 9821 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Removing barcodes with no covariate data\n",
      "View of AnnData object with n_obs × n_vars = 9821 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Structure details: AnnData object with n_obs × n_vars = 36016 × 32758 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'batch', 'batch_cov', 'ind_cov', 'well'\n",
      "    var: 'gene_ids-0-0-0', 'protein?-0-0-0', 'gene_ids-1-0-0', 'protein?-1-0-0', 'gene_ids-1-0', 'protein?-1-0', 'gene_ids-1', 'protein?-1'\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "Initial\n",
      "AnnData object with n_obs × n_vars = 20672 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "SNG confirmation via SNG- tag in BEST\n",
      "View of AnnData object with n_obs × n_vars = 15595 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Removing barcodes with no covariate data\n",
      "View of AnnData object with n_obs × n_vars = 14850 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Structure details: AnnData object with n_obs × n_vars = 50866 × 32758 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'batch', 'batch_cov', 'ind_cov', 'well'\n",
      "    var: 'gene_ids-0-0-0-0', 'protein?-0-0-0-0', 'gene_ids-1-0-0-0', 'protein?-1-0-0-0', 'gene_ids-1-0-0', 'protein?-1-0-0', 'gene_ids-1-0', 'protein?-1-0', 'gene_ids-1', 'protein?-1'\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "Initial\n",
      "AnnData object with n_obs × n_vars = 19143 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "SNG confirmation via SNG- tag in BEST\n",
      "View of AnnData object with n_obs × n_vars = 14987 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Removing barcodes with no covariate data\n",
      "View of AnnData object with n_obs × n_vars = 14295 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Structure details: AnnData object with n_obs × n_vars = 65161 × 32758 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'batch', 'batch_cov', 'ind_cov', 'well'\n",
      "    var: 'gene_ids-0-0-0-0-0', 'protein?-0-0-0-0-0', 'gene_ids-1-0-0-0-0', 'protein?-1-0-0-0-0', 'gene_ids-1-0-0-0', 'protein?-1-0-0-0', 'gene_ids-1-0-0', 'protein?-1-0-0', 'gene_ids-1-0', 'protein?-1-0', 'gene_ids-1', 'protein?-1'\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "Initial\n",
      "AnnData object with n_obs × n_vars = 19175 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "SNG confirmation via SNG- tag in BEST\n",
      "View of AnnData object with n_obs × n_vars = 14725 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Removing barcodes with no covariate data\n",
      "View of AnnData object with n_obs × n_vars = 14066 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Structure details: AnnData object with n_obs × n_vars = 79227 × 32758 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'batch', 'batch_cov', 'ind_cov', 'well'\n",
      "    var: 'gene_ids-0-0-0-0-0-0', 'protein?-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0', 'protein?-1-0-0-0-0-0', 'gene_ids-1-0-0-0-0', 'protein?-1-0-0-0-0', 'gene_ids-1-0-0-0', 'protein?-1-0-0-0', 'gene_ids-1-0-0', 'protein?-1-0-0', 'gene_ids-1-0', 'protein?-1-0', 'gene_ids-1', 'protein?-1'\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "Initial\n",
      "AnnData object with n_obs × n_vars = 20389 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "SNG confirmation via SNG- tag in BEST\n",
      "View of AnnData object with n_obs × n_vars = 15543 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Removing barcodes with no covariate data\n",
      "View of AnnData object with n_obs × n_vars = 14817 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Structure details: AnnData object with n_obs × n_vars = 94044 × 32758 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'batch', 'batch_cov', 'ind_cov', 'well'\n",
      "    var: 'gene_ids-0-0-0-0-0-0-0', 'protein?-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0', 'protein?-1-0-0-0-0-0', 'gene_ids-1-0-0-0-0', 'protein?-1-0-0-0-0', 'gene_ids-1-0-0-0', 'protein?-1-0-0-0', 'gene_ids-1-0-0', 'protein?-1-0-0', 'gene_ids-1-0', 'protein?-1-0', 'gene_ids-1', 'protein?-1'\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial\n",
      "AnnData object with n_obs × n_vars = 23259 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "SNG confirmation via SNG- tag in BEST\n",
      "View of AnnData object with n_obs × n_vars = 17449 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Removing barcodes with no covariate data\n",
      "View of AnnData object with n_obs × n_vars = 16725 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Structure details: AnnData object with n_obs × n_vars = 110769 × 32758 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'batch', 'batch_cov', 'ind_cov', 'well'\n",
      "    var: 'gene_ids-0-0-0-0-0-0-0-0', 'protein?-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0', 'protein?-1-0-0-0-0-0', 'gene_ids-1-0-0-0-0', 'protein?-1-0-0-0-0', 'gene_ids-1-0-0-0', 'protein?-1-0-0-0', 'gene_ids-1-0-0', 'protein?-1-0-0', 'gene_ids-1-0', 'protein?-1-0', 'gene_ids-1', 'protein?-1'\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "Initial\n",
      "AnnData object with n_obs × n_vars = 22993 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "SNG confirmation via SNG- tag in BEST\n",
      "View of AnnData object with n_obs × n_vars = 17178 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Removing barcodes with no covariate data\n",
      "View of AnnData object with n_obs × n_vars = 16452 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Structure details: AnnData object with n_obs × n_vars = 127221 × 32758 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'batch', 'batch_cov', 'ind_cov', 'well'\n",
      "    var: 'gene_ids-0-0-0-0-0-0-0-0-0', 'protein?-0-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0', 'protein?-1-0-0-0-0-0', 'gene_ids-1-0-0-0-0', 'protein?-1-0-0-0-0', 'gene_ids-1-0-0-0', 'protein?-1-0-0-0', 'gene_ids-1-0-0', 'protein?-1-0-0', 'gene_ids-1-0', 'protein?-1-0', 'gene_ids-1', 'protein?-1'\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "Initial\n",
      "AnnData object with n_obs × n_vars = 21345 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "SNG confirmation via SNG- tag in BEST\n",
      "View of AnnData object with n_obs × n_vars = 15656 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Removing barcodes with no covariate data\n",
      "View of AnnData object with n_obs × n_vars = 15000 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Structure details: AnnData object with n_obs × n_vars = 142221 × 32758 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'batch', 'batch_cov', 'ind_cov', 'well'\n",
      "    var: 'gene_ids-0-0-0-0-0-0-0-0-0-0', 'protein?-0-0-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0', 'protein?-1-0-0-0-0-0', 'gene_ids-1-0-0-0-0', 'protein?-1-0-0-0-0', 'gene_ids-1-0-0-0', 'protein?-1-0-0-0', 'gene_ids-1-0-0', 'protein?-1-0-0', 'gene_ids-1-0', 'protein?-1-0', 'gene_ids-1', 'protein?-1'\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.adt', '.mrna', '.demuxed', '.freemuxed', '.h5ad'].\n",
      "Only considering the two last: ['.freemuxed', '.h5ad'].\n",
      "Initial\n",
      "AnnData object with n_obs × n_vars = 20493 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "SNG confirmation via SNG- tag in BEST\n",
      "View of AnnData object with n_obs × n_vars = 15359 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Removing barcodes with no covariate data\n",
      "View of AnnData object with n_obs × n_vars = 14699 × 32758 \n",
      "    obs: 'N.SNP', 'BEST', 'mUMI', 'aUMI', 'NUM.SNPS', 'DROPLET.TYPE', 'BEST.GUESS', 'DOUBLETS'\n",
      "    var: 'gene_ids', 'protein?'\n",
      "Structure details: AnnData object with n_obs × n_vars = 156920 × 32758 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'batch', 'batch_cov', 'ind_cov', 'well'\n",
      "    var: 'gene_ids-0-0-0-0-0-0-0-0-0-0-0', 'protein?-0-0-0-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0', 'protein?-1-0-0-0-0-0', 'gene_ids-1-0-0-0-0', 'protein?-1-0-0-0-0', 'gene_ids-1-0-0-0', 'protein?-1-0-0-0', 'gene_ids-1-0-0', 'protein?-1-0-0', 'gene_ids-1-0', 'protein?-1-0', 'gene_ids-1', 'protein?-1'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:119: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "# Flare cohort files\n",
    "files = [\n",
    "'/ye/yelabstore2/george/Sasha_Flare_Studies/10xcount/combined.h5ads/again/1_1.adt.mrna.demuxed.freemuxed.h5ad',\n",
    "'/ye/yelabstore2/george/Sasha_Flare_Studies/10xcount/combined.h5ads/again/1_2.adt.mrna.demuxed.freemuxed.h5ad',\n",
    "'/ye/yelabstore2/george/Sasha_Flare_Studies/10xcount/combined.h5ads/again/2_1.adt.mrna.demuxed.freemuxed.h5ad',\n",
    "'/ye/yelabstore2/george/Sasha_Flare_Studies/10xcount/combined.h5ads/again/2_2.adt.mrna.demuxed.freemuxed.h5ad',\n",
    "'/ye/yelabstore2/george/Sasha_Flare_Studies/10xcount/combined.h5ads/again/3_1.adt.mrna.demuxed.freemuxed.h5ad',\n",
    "'/ye/yelabstore2/george/Sasha_Flare_Studies/10xcount/combined.h5ads/again/3_2.adt.mrna.demuxed.freemuxed.h5ad',\n",
    "'/ye/yelabstore2/george/Sasha_Flare_Studies/10xcount/combined.h5ads/again/3_3.adt.mrna.demuxed.freemuxed.h5ad',\n",
    "'/ye/yelabstore2/george/Sasha_Flare_Studies/10xcount/combined.h5ads/again/3_4.adt.mrna.demuxed.freemuxed.h5ad',\n",
    "'/ye/yelabstore2/george/Sasha_Flare_Studies/10xcount/combined.h5ads/again/4_1.adt.mrna.demuxed.freemuxed.h5ad',\n",
    "'/ye/yelabstore2/george/Sasha_Flare_Studies/10xcount/combined.h5ads/again/4_2.adt.mrna.demuxed.freemuxed.h5ad',\n",
    "'/ye/yelabstore2/george/Sasha_Flare_Studies/10xcount/combined.h5ads/again/4_3.adt.mrna.demuxed.freemuxed.h5ad',\n",
    "'/ye/yelabstore2/george/Sasha_Flare_Studies/10xcount/combined.h5ads/again/4_4.adt.mrna.demuxed.freemuxed.h5ad']\n",
    "flaregps = [\n",
    "'1_1', '1_2', '2_1', '2_2',\n",
    "'3_1', '3_2', '3_3', '3_4',\n",
    "'4_1', '4_2', '4_3', '4_4']\n",
    "DOUBLETS_dict = dict()\n",
    "# Concatenate files\n",
    "for ii in range(len(files)):\n",
    "    df = pd.read_excel('flare.covars.again.xlsx')\n",
    "    # This is crucial because Untreated and Treated were run in different runs and this is the only way to map it back.\n",
    "    df = df[df['RUN_WELL'] == flaregps[ii]]\n",
    "    if ii ==0:\n",
    "        adata = sc.read(files[ii])\n",
    "        print('Initial')\n",
    "        print(adata)\n",
    "        # Keep only singlets\n",
    "        doubletlist2 = [adata.obs['BEST'][i].split('-')[0] for i in range(len(adata.obs['BEST']))]\n",
    "        adata.obs['DOUBLETS'] = doubletlist2\n",
    "        adata = adata[adata.obs['DOUBLETS']=='SNG']\n",
    "        print('SNG confirmation via SNG- tag in BEST')\n",
    "\n",
    "        print(adata)\n",
    "        inds0 = np.unique(adata.obs['BEST'].tolist())\n",
    "        inds =[]\n",
    "        for i in range(len(inds0)):\n",
    "            inds.append('-'.join(inds0[i].split('-')[1:]))\n",
    "        inds = pd.Series(inds)\n",
    "        inds0 = pd.Series(inds0)\n",
    "        booleanlist = inds0[inds.isin(np.unique(df['ID'].tolist())).tolist()]\n",
    "        print('Removing barcodes with no covariate data')\n",
    "        adata = adata[adata.obs['BEST'].isin(booleanlist)]\n",
    "        print(adata)\n",
    "        \n",
    "        inds0 = np.unique(adata.obs['BEST'].tolist())\n",
    "        inds = []\n",
    "        for i in range(len(inds0)):\n",
    "            inds.append('-'.join(inds0[i].split('-')[1:]))\n",
    "        \n",
    "        # Preallocate space\n",
    "        adata.obs['BEST'] = adata.obs['BEST'].astype('object')\n",
    "        adata.obs['SOURCE'] = adata.obs['BEST']\n",
    "        adata.obs['STATUS'] = adata.obs['BEST']\n",
    "        adata.obs['TR_UNT'] = adata.obs['BEST']\n",
    "        adata.obs['ind_cov'] = adata.obs['BEST']\n",
    "        for xcv in range(len(inds)):\n",
    "            adata.obs['SOURCE'][adata.obs['BEST']==inds0[xcv]] = df['SOURCE'][df['ID'].isin([inds[xcv]])].values[0]\n",
    "            adata.obs['STATUS'][adata.obs['BEST']==inds0[xcv]] = df['STATUS'][df['ID'].isin([inds[xcv]])].values[0]\n",
    "            adata.obs['TR_UNT'][adata.obs['BEST']==inds0[xcv]] = df['TR/UNT'][df['ID'].isin([inds[xcv]])].values[0]\n",
    "            if df['ALIAS'][df['ID'].isin([inds[xcv]])].values[0] == 'FLARE2,3,or8':\n",
    "                adata.obs['ind_cov'][adata.obs['BEST']==inds0[xcv]] = df['ID'][df['ID'].isin([inds[xcv]])].values[0].split(',')[0]\n",
    "            else:\n",
    "                adata.obs['ind_cov'][adata.obs['BEST']==inds0[xcv]] = df['ALIAS'][df['ID'].isin([inds[xcv]])].values[0]\n",
    "           \n",
    "        del adata.obs['N.SNP']\n",
    "        del adata.obs['BEST']\n",
    "        del adata.obs['mUMI']\n",
    "        del adata.obs['aUMI']\n",
    "        del adata.obs['NUM.SNPS']\n",
    "        del adata.obs['DROPLET.TYPE']\n",
    "        del adata.obs['BEST.GUESS']\n",
    "        del adata.obs['DOUBLETS']\n",
    "        print(str('Structure details: ' + str(adata)))\n",
    "        adata.obs['well'] = np.repeat(flaregps[ii], len(adata.obs_names))\n",
    "        adata.obs['batch_cov'] = np.repeat(flaregps[ii][:-2], len(adata.obs_names))\n",
    "    else:\n",
    "        bdata = sc.read(files[ii])\n",
    "        print('Initial')\n",
    "        print(bdata)\n",
    "        # Keep only singlets\n",
    "        doubletlist2 = [bdata.obs['BEST'][i].split('-')[0] for i in range(len(bdata.obs['BEST']))]\n",
    "        bdata.obs['DOUBLETS'] = doubletlist2\n",
    "        bdata = bdata[bdata.obs['DOUBLETS']=='SNG']\n",
    "        print('SNG confirmation via SNG- tag in BEST')\n",
    "        \n",
    "        print(bdata)\n",
    "        inds0 = np.unique(bdata.obs['BEST'].tolist())\n",
    "        inds = []\n",
    "        for i in range(len(inds0)):\n",
    "            inds.append('-'.join(inds0[i].split('-')[1:]))\n",
    "        inds = pd.Series(inds)\n",
    "        inds0 = pd.Series(inds0)\n",
    "        booleanlist = inds0[inds.isin(np.unique(df['ID'].tolist())).tolist()]\n",
    "        print('Removing barcodes with no covariate data')\n",
    "        bdata = bdata[bdata.obs['BEST'].isin(booleanlist)]\n",
    "        print(bdata)\n",
    "        \n",
    "        inds0 = np.unique(bdata.obs['BEST'].tolist())\n",
    "        inds = []\n",
    "        for i in range(len(inds0)):\n",
    "            inds.append('-'.join(inds0[i].split('-')[1:]))\n",
    "        \n",
    "        # Preallocate space\n",
    "        bdata.obs['BEST'] = bdata.obs['BEST'].astype('object')\n",
    "        bdata.obs['SOURCE'] = bdata.obs['BEST']\n",
    "        bdata.obs['STATUS'] = bdata.obs['BEST']\n",
    "        bdata.obs['TR_UNT'] = bdata.obs['BEST']\n",
    "        bdata.obs['ind_cov'] = bdata.obs['BEST']\n",
    "        for xcv in range(len(inds)):\n",
    "            bdata.obs['SOURCE'][bdata.obs['BEST']==inds0[xcv]] = df['SOURCE'][df['ID'].isin([inds[xcv]])].values[0]\n",
    "            bdata.obs['STATUS'][bdata.obs['BEST']==inds0[xcv]] = df['STATUS'][df['ID'].isin([inds[xcv]])].values[0]\n",
    "            bdata.obs['TR_UNT'][bdata.obs['BEST']==inds0[xcv]] = df['TR/UNT'][df['ID'].isin([inds[xcv]])].values[0]\n",
    "            if df['ALIAS'][df['ID'].isin([inds[xcv]])].values[0] == 'FLARE2,3,or8':\n",
    "                bdata.obs['ind_cov'][bdata.obs['BEST']==inds0[xcv]] = df['ID'][df['ID'].isin([inds[xcv]])].values[0].split(',')[0]\n",
    "            else:    \n",
    "                bdata.obs['ind_cov'][bdata.obs['BEST']==inds0[xcv]] = df['ALIAS'][df['ID'].isin([inds[xcv]])].values[0]\n",
    "                               \n",
    "        del bdata.obs['N.SNP']\n",
    "        del bdata.obs['BEST']\n",
    "        del bdata.obs['mUMI']\n",
    "        del bdata.obs['aUMI']\n",
    "        del bdata.obs['NUM.SNPS']\n",
    "        del bdata.obs['DROPLET.TYPE']\n",
    "        del bdata.obs['BEST.GUESS']\n",
    "        del bdata.obs['DOUBLETS']\n",
    "        bdata.obs['well'] = np.repeat(flaregps[ii], len(bdata.obs_names))\n",
    "        bdata.obs['batch_cov'] = np.repeat(flaregps[ii][:-2], len(bdata.obs_names))\n",
    "        # Concatenate files.\n",
    "        adata = adata.concatenate(bdata)\n",
    "        print(str('Structure details: ' + str(adata)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# House Keeping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "# Fix this mislabeling.\n",
    "adata.obs.ind_cov = adata.obs.ind_cov.astype('object')\n",
    "adata.obs.ind_cov = adata.obs.ind_cov.astype('str')\n",
    "# Mislabeled by wrong excel label.\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='101'] = '1240'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='101_101'] = '1240'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='0_900805200_900805200'] = '1804_1804'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='0_904194200_904194200'] = '904194200_904194200'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='0_1586_1586'] = '1586_1586'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='0_1763_1763'] = '1763_1763'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='1772'] = 'FLARE018'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='2132DNA01'] = 'FLARE019'\n",
    "# Change to healthy to be more descriptive.\n",
    "adata.obs['TR_UNT'][adata.obs['TR_UNT']=='-'] = 'HEALTHY'\n",
    "# Remove '-' person...\n",
    "adata = adata[adata.obs['ind_cov']!='-']\n",
    "# 1333 is unknown. Remove.\n",
    "adata = adata[adata.obs['ind_cov']!='1333_1333']\n",
    "# 0_902320200_902320200 does not exist. Remove.\n",
    "adata = adata[adata.obs['ind_cov']!='0_902320200_902320200']\n",
    "print('Initial ind labels')\n",
    "ind_list = np.unique(adata.obs['ind_cov'].tolist())\n",
    "print(ind_list)\n",
    "# Simplify labels\n",
    "for ii in range(len(adata.obs['ind_cov'])):\n",
    "    if adata.obs['ind_cov'][ii].startswith('0_'):\n",
    "        adata.obs['ind_cov'][ii] = '_'.join(adata.obs['ind_cov'][ii].split('_')[1:])\n",
    "print('Simplified labels')        \n",
    "ind_list = np.unique(adata.obs['ind_cov'].tolist())\n",
    "print(ind_list)\n",
    "# Remove individuals with barcode counts less than 100\n",
    "for ii in range(len(ind_list)):\n",
    "    if np.sum(adata.obs['ind_cov']==ind_list[ii]) < 100:\n",
    "        adata = adata[adata.obs['ind_cov']!=ind_list[ii]]\n",
    "    else:\n",
    "        continue\n",
    "ind_list = np.unique(adata.obs['ind_cov'].tolist())\n",
    "\n",
    "# From runs 1 and 2 these are labeled as healthy from CLUES but they are treated CLUES samples and not true healthies.\n",
    "adata.obs['TR_UNT'] = adata.obs['TR_UNT'].astype('object')\n",
    "adata.obs.loc[adata.obs.ind_cov == \"1754_1754\",'TR_UNT'] = \"TR\"\n",
    "adata.obs.loc[adata.obs.ind_cov == \"1771_1771\",'TR_UNT'] = \"TR\"\n",
    "adata.obs.loc[adata.obs.ind_cov == \"1791_1791\",'TR_UNT'] = \"TR\"\n",
    "adata.obs.loc[adata.obs.ind_cov == \"900805200_900805200\",'TR_UNT'] = \"TR\"\n",
    "adata.obs.loc[adata.obs.ind_cov == \"1240\",'TR_UNT'] = \"TR\"\n",
    "adata.obs.loc[adata.obs.ind_cov == \"1804_1804\",'TR_UNT'] = \"TR\"\n",
    "\n",
    "# Fix even more mislabeling ...\n",
    "adata.obs['TR_UNT'] = adata.obs['TR_UNT'].astype('str')\n",
    "# 904194200_904194200 #\n",
    "adata.obs['TR_UNT'][(adata.obs['TR_UNT']=='Unknown (1 TR, 2 UNT)') & (adata.obs['ind_cov']=='904194200_904194200')] = 'UNT'\n",
    "adata.obs['TR_UNT'][(adata.obs['TR_UNT']=='Unknown (2 TR, 1 UNT)') & (adata.obs['ind_cov']=='904194200_904194200')] = 'TR'\n",
    "# 1586_1586 #\n",
    "adata.obs['TR_UNT'][(adata.obs['TR_UNT']=='Unknown (1 TR, 2 UNT)') & (adata.obs['ind_cov']=='1586_1586')] = 'TR'\n",
    "adata.obs['TR_UNT'][(adata.obs['TR_UNT']=='Unknown (2 TR, 1 UNT)') & (adata.obs['ind_cov']=='1586_1586')] = 'UNT'\n",
    "# 1763_1763 # TR in Unknown (2 TR, 1 UNT)\n",
    "adata.obs['TR_UNT'][(adata.obs['TR_UNT']=='Unknown (1 TR, 2 UNT)') & (adata.obs['ind_cov']=='1763_1763')] = 'UNT'\n",
    "adata.obs['TR_UNT'][(adata.obs['TR_UNT']=='Unknown (2 TR, 1 UNT)') & (adata.obs['ind_cov']=='1763_1763')] = 'TR'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add female and broad covariates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'SOURCE' as categorical\n",
      "... storing 'STATUS' as categorical\n",
      "... storing 'TR_UNT' as categorical\n",
      "... storing 'batch_cov' as categorical\n",
      "... storing 'ind_cov' as categorical\n",
      "... storing 'well' as categorical\n",
      "... storing 'pop_cov' as categorical\n",
      "... storing 'protein?-0-0-0-0-0-0-0-0-0-0-0' as categorical\n",
      "... storing 'protein?-1-0-0-0-0-0-0-0-0-0-0' as categorical\n",
      "... storing 'protein?-1-0-0-0-0-0-0-0-0-0' as categorical\n",
      "... storing 'protein?-1-0-0-0-0-0-0-0-0' as categorical\n",
      "... storing 'protein?-1-0-0-0-0-0-0-0' as categorical\n",
      "... storing 'protein?-1-0-0-0-0-0-0' as categorical\n",
      "... storing 'protein?-1-0-0-0-0-0' as categorical\n",
      "... storing 'protein?-1-0-0-0-0' as categorical\n",
      "... storing 'protein?-1-0-0-0' as categorical\n",
      "... storing 'protein?-1-0-0' as categorical\n",
      "... storing 'protein?-1-0' as categorical\n",
      "... storing 'protein?-1' as categorical\n"
     ]
    }
   ],
   "source": [
    "%%capture\n",
    "# preallocate space\n",
    "adata.obs['Female'] = adata.obs.ind_cov.astype('object')\n",
    "adata.obs['pop_cov'] = adata.obs.ind_cov.astype('object')\n",
    "genderdict = {'1240':1, '1716_1716':1, '1804_1804':1,'1754_1754':1, '1731_1731':1, '1763_1763':1, \n",
    "'1771_1771':1, '1791_1791':1, '1892_1892':1,'904236200_904236200':1, '1243_1243':1, 'FLARE004':1, 'FLARE005':1, 'FLARE006':1, 'FLARE007':1,\n",
    " 'FLARE009':1, 'FLARE011':1, 'FLARE013':1, '904194200_904194200':1, '1586_1586':1, '1763_1763':1,'FLARE014':1, 'FLARE015':1,\n",
    " 'FLARE016':1, 'FLARE017':1, 'FLARE018':1, 'FLARE019':0, 'FLARE020':1,'FLARE2,3,or8':1, 'HC-002':0, 'HC-014':1,\n",
    " 'HC-020':1, 'HC-501':1, 'HC-502':1, 'HC-503':1, 'HC-504':1, 'HC-506':1,\n",
    " 'IGTB1290':1, 'IGTB1372':1, 'IGTB1506':1, 'IGTB1921':1, 'IGTB195':1, 'IGTB256':1, 'IGTB469':1,\n",
    " 'IGTB508':1, 'IGTB514':1, 'IGTB826':1}\n",
    "\n",
    "popdict = {'1240':'ASIAN', '1716_1716':'WHITE', '1804_1804':'ASIAN','1754_1754':'ASIAN', '1731_1731':'WHITE', '1763_1763':'ASIAN', \n",
    "'1771_1771':'ASIAN', '904194200_904194200':'WHITE', '1586_1586':\"ASIAN\", '1763_1763':'ASIAN' ,'1791_1791':'ASIAN', '1892_1892':'WHITE', '904236200_904236200':'ASIAN', '1243_1243':'WHITE', \n",
    "           'FLARE004':'ASIAN', 'FLARE005':'WHITE', 'FLARE006':'WHITE', 'FLARE007':'ASIAN',\n",
    " 'FLARE009':'HISPANIC', 'FLARE011':'ASIAN', 'FLARE013':'BLACK', 'FLARE014':'BLACK', 'FLARE015':'ASIAN',\n",
    " 'FLARE016':'ASIAN', 'FLARE017':'BLACK', 'FLARE018':'ASIAN', 'FLARE019':'HISPANIC', 'FLARE020':'WHITE','FLARE2,3,or8':'UNKNOWN', 'HC-002':'WHITE', 'HC-014':'ASIAN',\n",
    " 'HC-020':'UNKNOWN', 'HC-501':'ASIAN', 'HC-502':'WHITE', 'HC-503':'WHITE', 'HC-504':'WHITE', 'HC-506':'WHITE',\n",
    " 'IGTB1290':'WHITE', 'IGTB1372':'WHITE', 'IGTB1506':'WHITE', 'IGTB1921':'WHITE', 'IGTB195':'WHITE', 'IGTB256':'WHITE', 'IGTB469':'WHITE',\n",
    " 'IGTB508':'WHITE', 'IGTB514':'WHITE', 'IGTB826':'WHITE'}\n",
    "\n",
    "for key in genderdict.keys():\n",
    "    adata.obs.Female[adata.obs.ind_cov==key] = genderdict[key]\n",
    "    \n",
    "for key in popdict.keys():\n",
    "    adata.obs.pop_cov[adata.obs.ind_cov==key] = popdict[key]\n",
    "    \n",
    "# Preallocate space\n",
    "adata.obs['Broad'] = adata.obs.SOURCE.astype('object')\n",
    "# if immvar is source, assign 1.\n",
    "adata.obs['Broad'][adata.obs.SOURCE == 'IMMVAR'] = 1\n",
    "adata.obs['Broad'][adata.obs.SOURCE != 'IMMVAR'] = 0\n",
    "\n",
    "adata.obs['Female'] = adata.obs['Female'].astype(dtype=np.float32)\n",
    "adata.obs['Broad'] = adata.obs['Broad'].astype(dtype=np.float32)\n",
    "adata.write(savepath, compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fix a bug and move Ab data out of gene matrix and take Center-log-ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Your filename has more than two extensions: ['.7', '.flareAb', '.h5ad'].\n",
      "Only considering the two last: ['.flareAb', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.7', '.flareAb', '.h5ad'].\n",
      "Only considering the two last: ['.flareAb', '.h5ad'].\n",
      "abmatrix shape: (156903, 20)\n",
      "Structure details: View of AnnData object with n_obs × n_vars = 156903 × 32738 \n",
      "    obs: 'SOURCE', 'STATUS', 'TR_UNT', 'batch', 'batch_cov', 'ind_cov', 'well', 'Female', 'pop_cov', 'Broad', 'CD38|CD38|j95-01|pAbO', 'CD95|FAS|j95-02|pAbO', 'CD28|CD28|j95-07|pAbO', 'CD27|CD27|j95-08|pAbO', 'CD197|CCR7|j95-09|pAbO', 'CD19|CD19|j95-10|pAbO', 'CD161|KLRB1|j95-11|pAbO', 'CD183|CXCR3|j95-12|pAbO', 'CD4|CD4|j95-14|pAbO', 'CD3|CD3E|j95-16|pAbO', 'CD196|CCR6|j95-17|pAbO', 'HLA-DR|CD74|j95-18|pAbO', 'CD45RO|PTPRC|j95-19|pAbO', 'CD14|CD14|j95-20|pAbO', 'CD194|CCR4|j95-21|pAbO', 'CD185|CXCR5|j95-22|pAbO', 'CD25|IL2RA|j95-24|pAbO', 'CD8|CD8A|j95-25|pAbO', 'CD127|IL7R|j95-27|pAbO', 'CD45RA|PTPRC|j95-28|pAbO'\n",
      "    var: 'gene_ids-0-0-0-0-0-0-0-0-0-0-0', 'protein?-0-0-0-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0-0', 'protein?-1-0-0-0-0-0-0', 'gene_ids-1-0-0-0-0-0', 'protein?-1-0-0-0-0-0', 'gene_ids-1-0-0-0-0', 'protein?-1-0-0-0-0', 'gene_ids-1-0-0-0', 'protein?-1-0-0-0', 'gene_ids-1-0-0', 'protein?-1-0-0', 'gene_ids-1-0', 'protein?-1-0', 'gene_ids-1', 'protein?-1'\n"
     ]
    }
   ],
   "source": [
    "from skbio.stats.composition import clr\n",
    "adata = sc.read(savepath)\n",
    "# Fix gene endings\n",
    "genelist = adata.var_names.tolist()\n",
    "genelist = [gn[:-2] for gn in genelist]\n",
    "adata.var_names = genelist\n",
    "# Find antibodies\n",
    "ab_names = [ab for ab in genelist if 'pAbO' in ab]\n",
    "# Find antibody index\n",
    "ab_index = [genelist.index(ab) for ab in ab_names]\n",
    "# Make new adata structure with just the antibody results\n",
    "ABdata = adata[:, ab_index]\n",
    "mat = csr_matrix(ABdata.X)\n",
    "mat = mat.todense()\n",
    "matrix = mat\n",
    "print(str('abmatrix shape: ' + str(np.shape(matrix))))\n",
    "# Make antibody into observations\n",
    "for ii in range(len(ab_names)):\n",
    "    LIST = np.asarray(matrix[:,ii])\n",
    "    adata.obs[ab_names[ii]] = LIST\n",
    "    \n",
    "# Remove antibodies from gene list\n",
    "abbool = [not ab for ab in adata.var_names.isin(ab_names)]\n",
    "adata = adata[:, abbool]\n",
    "print(str('Structure details: ' + str(adata)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove all var lists to clean up structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for key in list(adata.var.keys()):\n",
    "    del adata.var[key]\n",
    "adata.obs['disease_cov'] = adata.obs['TR_UNT']\n",
    "del adata.obs['batch']\n",
    "del adata.obs['STATUS']\n",
    "del adata.obs['TR_UNT']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process flare cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Data structure details: AnnData object with n_obs × n_vars = 156903 × 32738 \n",
      "    obs: 'SOURCE', 'batch_cov', 'ind_cov', 'well', 'Female', 'pop_cov', 'Broad', 'CD38|CD38|j95-01|pAbO', 'CD95|FAS|j95-02|pAbO', 'CD28|CD28|j95-07|pAbO', 'CD27|CD27|j95-08|pAbO', 'CD197|CCR7|j95-09|pAbO', 'CD19|CD19|j95-10|pAbO', 'CD161|KLRB1|j95-11|pAbO', 'CD183|CXCR3|j95-12|pAbO', 'CD4|CD4|j95-14|pAbO', 'CD3|CD3E|j95-16|pAbO', 'CD196|CCR6|j95-17|pAbO', 'HLA-DR|CD74|j95-18|pAbO', 'CD45RO|PTPRC|j95-19|pAbO', 'CD14|CD14|j95-20|pAbO', 'CD194|CCR4|j95-21|pAbO', 'CD185|CXCR5|j95-22|pAbO', 'CD25|IL2RA|j95-24|pAbO', 'CD8|CD8A|j95-25|pAbO', 'CD127|IL7R|j95-27|pAbO', 'CD45RA|PTPRC|j95-28|pAbO', 'disease_cov'\n",
      "INFO:root:Data structure details: AnnData object with n_obs × n_vars = 156903 × 32738 \n",
      "    obs: 'SOURCE', 'batch_cov', 'ind_cov', 'well', 'Female', 'pop_cov', 'Broad', 'CD38|CD38|j95-01|pAbO', 'CD95|FAS|j95-02|pAbO', 'CD28|CD28|j95-07|pAbO', 'CD27|CD27|j95-08|pAbO', 'CD197|CCR7|j95-09|pAbO', 'CD19|CD19|j95-10|pAbO', 'CD161|KLRB1|j95-11|pAbO', 'CD183|CXCR3|j95-12|pAbO', 'CD4|CD4|j95-14|pAbO', 'CD3|CD3E|j95-16|pAbO', 'CD196|CCR6|j95-17|pAbO', 'HLA-DR|CD74|j95-18|pAbO', 'CD45RO|PTPRC|j95-19|pAbO', 'CD14|CD14|j95-20|pAbO', 'CD194|CCR4|j95-21|pAbO', 'CD185|CXCR5|j95-22|pAbO', 'CD25|IL2RA|j95-24|pAbO', 'CD8|CD8A|j95-25|pAbO', 'CD127|IL7R|j95-27|pAbO', 'CD45RA|PTPRC|j95-28|pAbO', 'disease_cov'\n",
      "INFO:root:Mito genes: ['MT-ND1', 'MT-ND2', 'MT-CO1', 'MT-CO2', 'MT-ATP8', 'MT-ATP6', 'MT-CO3', 'MT-ND3', 'MT-ND4L', 'MT-ND4', 'MT-ND5', 'MT-ND6', 'MT-CYB']\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:35: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "INFO:root:Saving log(counts)+1 in .raw\n",
      "INFO:root:Filtering cells\n",
      "INFO:root:Data structure details: View of AnnData object with n_obs × n_vars = 154820 × 32738 \n",
      "    obs: 'SOURCE', 'batch_cov', 'ind_cov', 'well', 'Female', 'pop_cov', 'Broad', 'CD38|CD38|j95-01|pAbO', 'CD95|FAS|j95-02|pAbO', 'CD28|CD28|j95-07|pAbO', 'CD27|CD27|j95-08|pAbO', 'CD197|CCR7|j95-09|pAbO', 'CD19|CD19|j95-10|pAbO', 'CD161|KLRB1|j95-11|pAbO', 'CD183|CXCR3|j95-12|pAbO', 'CD4|CD4|j95-14|pAbO', 'CD3|CD3E|j95-16|pAbO', 'CD196|CCR6|j95-17|pAbO', 'HLA-DR|CD74|j95-18|pAbO', 'CD45RO|PTPRC|j95-19|pAbO', 'CD14|CD14|j95-20|pAbO', 'CD194|CCR4|j95-21|pAbO', 'CD185|CXCR5|j95-22|pAbO', 'CD25|IL2RA|j95-24|pAbO', 'CD8|CD8A|j95-25|pAbO', 'CD127|IL7R|j95-27|pAbO', 'CD45RA|PTPRC|j95-28|pAbO', 'disease_cov', 'percent_mito', 'n_counts'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtered out 865 cells that have more than  2500 genes expressed\n",
      "filtered out 17883 genes that are detected in less than 100 cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Data structure details: AnnData object with n_obs × n_vars = 153955 × 14855 \n",
      "    obs: 'SOURCE', 'batch_cov', 'ind_cov', 'well', 'Female', 'pop_cov', 'Broad', 'CD38|CD38|j95-01|pAbO', 'CD95|FAS|j95-02|pAbO', 'CD28|CD28|j95-07|pAbO', 'CD27|CD27|j95-08|pAbO', 'CD197|CCR7|j95-09|pAbO', 'CD19|CD19|j95-10|pAbO', 'CD161|KLRB1|j95-11|pAbO', 'CD183|CXCR3|j95-12|pAbO', 'CD4|CD4|j95-14|pAbO', 'CD3|CD3E|j95-16|pAbO', 'CD196|CCR6|j95-17|pAbO', 'HLA-DR|CD74|j95-18|pAbO', 'CD45RO|PTPRC|j95-19|pAbO', 'CD14|CD14|j95-20|pAbO', 'CD194|CCR4|j95-21|pAbO', 'CD185|CXCR5|j95-22|pAbO', 'CD25|IL2RA|j95-24|pAbO', 'CD8|CD8A|j95-25|pAbO', 'CD127|IL7R|j95-27|pAbO', 'CD45RA|PTPRC|j95-28|pAbO', 'disease_cov', 'percent_mito', 'n_counts', 'n_genes'\n",
      "    var: 'n_cells'\n",
      "INFO:root:Saving raw counts\n",
      "INFO:root:Normalizing total counts to 10,000\n",
      "INFO:root:Log transforming data\n",
      "INFO:root:Making .obs into categories\n",
      "... storing 'ind_cov' as categorical\n",
      "INFO:root:Filtering genes\n",
      "INFO:root:Data structure details: View of AnnData object with n_obs × n_vars = 153955 × 1164 \n",
      "    obs: 'SOURCE', 'batch_cov', 'ind_cov', 'well', 'Female', 'pop_cov', 'Broad', 'CD38|CD38|j95-01|pAbO', 'CD95|FAS|j95-02|pAbO', 'CD28|CD28|j95-07|pAbO', 'CD27|CD27|j95-08|pAbO', 'CD197|CCR7|j95-09|pAbO', 'CD19|CD19|j95-10|pAbO', 'CD161|KLRB1|j95-11|pAbO', 'CD183|CXCR3|j95-12|pAbO', 'CD4|CD4|j95-14|pAbO', 'CD3|CD3E|j95-16|pAbO', 'CD196|CCR6|j95-17|pAbO', 'HLA-DR|CD74|j95-18|pAbO', 'CD45RO|PTPRC|j95-19|pAbO', 'CD14|CD14|j95-20|pAbO', 'CD194|CCR4|j95-21|pAbO', 'CD185|CXCR5|j95-22|pAbO', 'CD25|IL2RA|j95-24|pAbO', 'CD8|CD8A|j95-25|pAbO', 'CD127|IL7R|j95-27|pAbO', 'CD45RA|PTPRC|j95-28|pAbO', 'disease_cov', 'percent_mito', 'n_counts', 'n_genes', 'PF4', 'SDPR', 'GNG11', 'PPBP'\n",
      "    var: 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
      "    uns: 'barcodes', 'genes', 'raw_counts'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing PCA with n_comps = 50\n",
      "computing PCA on highly variable genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:0.03289466194222416\n",
      "INFO:root:-0.024256113962236408\n",
      "INFO:root:0.04943114416692915\n",
      "INFO:root:0.6864686745989058\n",
      "INFO:root:3\n",
      "INFO:root:[[1.         0.68646867]\n",
      " [0.68646867 1.        ]]\n",
      "INFO:root:-0.019245743459576612\n",
      "INFO:root:-0.07126194870209629\n",
      "INFO:root:-0.021853658160646584\n",
      "INFO:root:-0.028917875408797835\n",
      "INFO:root:-0.019351945929253878\n",
      "INFO:root:-0.013843134135526694\n",
      "INFO:root:-0.017616936052614027\n",
      "INFO:root:-0.010341023437602485\n",
      "INFO:root:-0.022153444329794957\n",
      "INFO:root:-0.03416881233664512\n",
      "INFO:root:0.021039658915079317\n",
      "INFO:root:0.0014076111880392391\n",
      "INFO:root:-0.21526329380911205\n",
      "INFO:root:16\n",
      "INFO:root:[[ 1.         -0.21526329]\n",
      " [-0.21526329  1.        ]]\n",
      "INFO:root:0.02776355131900495\n",
      "INFO:root:0.012410002207283118\n",
      "INFO:root:0.012528772553391196\n",
      "INFO:root:0.023349977152981594\n",
      "INFO:root:-0.018581663064055225\n",
      "INFO:root:-0.006097165326742777\n",
      "INFO:root:0.006777747781682785\n",
      "INFO:root:-0.023183382807453815\n",
      "INFO:root:-0.02214776765600093\n",
      "INFO:root:-0.010306573984465384\n",
      "INFO:root:-0.0021879462225706018\n",
      "INFO:root:-0.0027603877389908676\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    finished (0:00:24.97)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:0.0303961759022956\n",
      "INFO:root:0.0039299453934165405\n",
      "INFO:root:-0.009046957535637591\n",
      "INFO:root:0.006116870848903306\n",
      "INFO:root:-0.020344405017228045\n",
      "INFO:root:0.0026080122407125266\n",
      "INFO:root:0.022142741432317287\n",
      "INFO:root:-0.00907602048362734\n",
      "INFO:root:0.01847156738948212\n",
      "INFO:root:0.01073107308393658\n",
      "INFO:root:-0.015916138749727516\n",
      "INFO:root:0.016880993756733145\n",
      "INFO:root:-0.07625616157868897\n",
      "INFO:root:-0.021244323027525887\n",
      "INFO:root:0.03185191264050187\n",
      "INFO:root:-0.056002082072584725\n",
      "INFO:root:0.03410840937138778\n",
      "INFO:root:0.0012239835480657702\n",
      "INFO:root:-0.0035522705143441577\n",
      "INFO:root:0.002799744598944874\n",
      "INFO:root:-0.019563912855874208\n",
      "INFO:root:Regressing out total nUMIs, percentage mitochondrial UMIs, gender, platelet signature and pool\n",
      "INFO:root:regressors: ['n_counts', 'percent_mito', 'Female', 'PC3', 'PC16']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "regressing out ['n_counts', 'percent_mito', 'Female', 'PC3', 'PC16']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Scaling expression data\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    finished (0:02:15.59)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Running combat\n",
      "found 4 batches\n",
      "found 0 numerical covariates...\n",
      "found 2 categorical variables:\tdisease_cov[T.TR], disease_cov[T.UNT]\n",
      "Standardizing Data across genes.\n",
      "Fitting L/S model and finding priors\n",
      "Finding parametric adjustments\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adjusting data\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Data structure details: AnnData object with n_obs × n_vars = 153955 × 1164 \n",
      "    obs: 'SOURCE', 'batch_cov', 'ind_cov', 'well', 'Female', 'pop_cov', 'Broad', 'CD38|CD38|j95-01|pAbO', 'CD95|FAS|j95-02|pAbO', 'CD28|CD28|j95-07|pAbO', 'CD27|CD27|j95-08|pAbO', 'CD197|CCR7|j95-09|pAbO', 'CD19|CD19|j95-10|pAbO', 'CD161|KLRB1|j95-11|pAbO', 'CD183|CXCR3|j95-12|pAbO', 'CD4|CD4|j95-14|pAbO', 'CD3|CD3E|j95-16|pAbO', 'CD196|CCR6|j95-17|pAbO', 'HLA-DR|CD74|j95-18|pAbO', 'CD45RO|PTPRC|j95-19|pAbO', 'CD14|CD14|j95-20|pAbO', 'CD194|CCR4|j95-21|pAbO', 'CD185|CXCR5|j95-22|pAbO', 'CD25|IL2RA|j95-24|pAbO', 'CD8|CD8A|j95-25|pAbO', 'CD127|IL7R|j95-27|pAbO', 'CD45RA|PTPRC|j95-28|pAbO', 'disease_cov', 'percent_mito', 'n_counts', 'n_genes', 'PF4', 'SDPR', 'GNG11', 'PPBP', 'PC3', 'PC16'\n",
      "    var: 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
      "    uns: 'barcodes', 'genes', 'raw_counts', 'pca'\n",
      "    obsm: 'X_pca'\n",
      "    varm: 'PCs'\n",
      "INFO:root:Removing samples that were processed twice...\n",
      "INFO:root:Data structure details: View of AnnData object with n_obs × n_vars = 153955 × 1164 \n",
      "    obs: 'SOURCE', 'batch_cov', 'ind_cov', 'well', 'Female', 'pop_cov', 'Broad', 'CD38|CD38|j95-01|pAbO', 'CD95|FAS|j95-02|pAbO', 'CD28|CD28|j95-07|pAbO', 'CD27|CD27|j95-08|pAbO', 'CD197|CCR7|j95-09|pAbO', 'CD19|CD19|j95-10|pAbO', 'CD161|KLRB1|j95-11|pAbO', 'CD183|CXCR3|j95-12|pAbO', 'CD4|CD4|j95-14|pAbO', 'CD3|CD3E|j95-16|pAbO', 'CD196|CCR6|j95-17|pAbO', 'HLA-DR|CD74|j95-18|pAbO', 'CD45RO|PTPRC|j95-19|pAbO', 'CD14|CD14|j95-20|pAbO', 'CD194|CCR4|j95-21|pAbO', 'CD185|CXCR5|j95-22|pAbO', 'CD25|IL2RA|j95-24|pAbO', 'CD8|CD8A|j95-25|pAbO', 'CD127|IL7R|j95-27|pAbO', 'CD45RA|PTPRC|j95-28|pAbO', 'disease_cov', 'percent_mito', 'n_counts', 'n_genes', 'PF4', 'SDPR', 'GNG11', 'PPBP', 'PC3', 'PC16'\n",
      "    var: 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
      "    uns: 'barcodes', 'genes', 'raw_counts', 'pca'\n",
      "    obsm: 'X_pca'\n",
      "    varm: 'PCs'\n",
      "Up to anndata 0.6.12, `.copy()` cast a non-'float32' `.X` to 'float32'. Now, the dtype 'float64' is maintained. \n",
      "INFO:root:Saving processed data\n",
      "INFO:root:Structure details: AnnData object with n_obs × n_vars = 153955 × 1164 \n",
      "    obs: 'SOURCE', 'batch_cov', 'ind_cov', 'well', 'Female', 'pop_cov', 'Broad', 'CD38|CD38|j95-01|pAbO', 'CD95|FAS|j95-02|pAbO', 'CD28|CD28|j95-07|pAbO', 'CD27|CD27|j95-08|pAbO', 'CD197|CCR7|j95-09|pAbO', 'CD19|CD19|j95-10|pAbO', 'CD161|KLRB1|j95-11|pAbO', 'CD183|CXCR3|j95-12|pAbO', 'CD4|CD4|j95-14|pAbO', 'CD3|CD3E|j95-16|pAbO', 'CD196|CCR6|j95-17|pAbO', 'HLA-DR|CD74|j95-18|pAbO', 'CD45RO|PTPRC|j95-19|pAbO', 'CD14|CD14|j95-20|pAbO', 'CD194|CCR4|j95-21|pAbO', 'CD185|CXCR5|j95-22|pAbO', 'CD25|IL2RA|j95-24|pAbO', 'CD8|CD8A|j95-25|pAbO', 'CD127|IL7R|j95-27|pAbO', 'CD45RA|PTPRC|j95-28|pAbO', 'disease_cov', 'percent_mito', 'n_counts', 'n_genes', 'PF4', 'SDPR', 'GNG11', 'PPBP', 'PC3', 'PC16'\n",
      "    uns: 'barcodes', 'genes', 'raw_counts', 'pca', 'total_pbmcs'\n",
      "    obsm: 'X_pca'\n",
      "    varm: 'PCs'\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:133: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:134: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:135: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:136: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:137: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:138: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:139: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:145: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:146: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:147: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/netapp/home/richardkperez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:148: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "... storing 'ind_cov' as categorical\n",
      "... storing 'disease_cov' as categorical\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import scanpy.api as sc\n",
    "from scipy.sparse import csr_matrix\n",
    "import logging\n",
    "import doubletdetection\n",
    "import pandas as pd\n",
    "import combat2\n",
    "import combat\n",
    "import patsy\n",
    "\n",
    "##################\n",
    "# Configure file #\n",
    "##################\n",
    "sc.settings.verbosity = 2\n",
    "sc.settings.autoshow = False\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "adata.obs['well'] = adata.obs['well'].astype('category')\n",
    "adata.var_names_make_unique()\n",
    "logging.info(str('Data structure details: ' + str(adata)))\n",
    "mat = csr_matrix(adata.X)\n",
    "logging.info(str('Data structure details: ' + str(adata)))\n",
    "# Extract list of genes\n",
    "genelist = adata.var_names.tolist()\n",
    "# Find mitochondrial genes\n",
    "mito_genes_names = [gn for gn in genelist if gn.startswith('MT-')]\n",
    "logging.info(str('Mito genes: ' + str(mito_genes_names)))\n",
    "# Find indices of mitochondrial genes\n",
    "mito_genes = [genelist.index(gn) for gn in mito_genes_names]\n",
    "# For each cell compute fraction of counts in mito genes vs. all genes\n",
    "adata.obs['percent_mito'] = np.ravel(np.sum(adata[:, mito_genes].X, axis=1)) / np.ravel(np.sum(adata.X, axis=1))\n",
    "# Add the total counts per cell as observations-annotation to adata\n",
    "adata.obs['n_counts'] = np.ravel(adata.X.sum(axis=1))\n",
    "# Fix name to make it compatible with clinical variates\n",
    "adata.obs['ind_cov'] = adata.obs['ind_cov'].astype('object')\n",
    "adata.obs['ind_cov'][adata.obs['ind_cov'].isin(['1221'])] = '1221_1221'\n",
    "adata.obs['ind_cov'][adata.obs['ind_cov'].isin(['1251'])] = '1251_1251'\n",
    "adata.obs['ind_cov'][adata.obs['ind_cov'].isin(['1891'])] = '1891_1891'\n",
    "indlist = np.unique(np.asarray(adata.obs['ind_cov'].tolist()))\n",
    "# Get list of batches\n",
    "batch_list = adata.obs['batch_cov'].tolist()\n",
    "logging.info('Saving log(counts)+1 in .raw')\n",
    "adata.raw = sc.pp.log1p(adata, copy=True)\n",
    "logging.info('Filtering cells')\n",
    "# Filter cells that have more than 10% of counts coming from mitochondrial genes.\n",
    "adata = adata[adata.obs['percent_mito'] < 0.10]\n",
    "logging.info(str('Data structure details: ' + str(adata)))\n",
    "# Filter cells with abnormally low gene counts, high gene counts.\n",
    "sc.pp.filter_cells(adata, min_genes=100)\n",
    "sc.pp.filter_cells(adata, max_genes=2500)\n",
    "sc.pp.filter_genes(adata, min_cells=100)\n",
    "logging.info(str('Data structure details: ' + str(adata)))\n",
    "logging.info('Saving raw counts')\n",
    "adata.uns['barcodes'] = adata.obs_names.tolist()\n",
    "adata.uns['genes'] = adata.var_names.tolist()\n",
    "adata.uns['raw_counts'] = adata.X\n",
    "logging.info('Normalizing total counts to 10,000')\n",
    "sc.pp.normalize_per_cell(adata, counts_per_cell_after=1e4)\n",
    "logging.info('Log transforming data')\n",
    "sc.pp.log1p(adata)\n",
    "# Add platelet genes for purposes of regressing out their signature\n",
    "adata.obs['PF4'] = adata.raw[:,\"PF4\"].X\n",
    "adata.obs['SDPR'] = adata.raw[:,\"SDPR\"].X\n",
    "adata.obs['GNG11'] = adata.raw[:,\"GNG11\"].X\n",
    "adata.obs['PPBP'] = adata.raw[:,\"PPBP\"].X\n",
    "logging.info('Making .obs into categories')\n",
    "adata.strings_to_categoricals()\n",
    "logging.info('Filtering genes')\n",
    "sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5, inplace=True)\n",
    "adata = adata[:, adata.var['highly_variable']]\n",
    "logging.info(str('Data structure details: ' + str(adata)))\n",
    "sc.pp.scale(adata, max_value=10)\n",
    "# Compute PCA\n",
    "sc.pp.pca(adata, random_state=1, svd_solver='arpack')\n",
    "# Removing the two components most highly correlated with platelet counts\n",
    "to_remove = pd.Series()\n",
    "for ii in range(len(adata.obsm.X_pca[1,:])) :\n",
    "    logging.info(str(np.corrcoef(adata.obsm.X_pca[:,ii], adata.obs[\"PF4\"])[0,1]))\n",
    "    if(np.absolute(np.corrcoef(adata.obsm.X_pca[:,ii], adata.obs[\"PF4\"])[0,1]) > 0.1):\n",
    "        adata.obs[\"PC\"+str(ii)] = adata.obsm.X_pca[:,ii];\n",
    "        to_remove = to_remove.append(pd.Series([\"PC\"+str(ii)]))\n",
    "        logging.info(str(ii))\n",
    "        logging.info(str(np.corrcoef(adata.obsm.X_pca[:,ii], adata.obs[\"PF4\"])))\n",
    "\n",
    "regressors = ['n_counts', 'percent_mito', 'Female']+to_remove.tolist();\n",
    "logging.info('Regressing out total nUMIs, percentage mitochondrial UMIs, gender, platelet signature and pool')\n",
    "logging.info('regressors: '+str(regressors))\n",
    "sc.pp.regress_out(adata, regressors)\n",
    "logging.info('Scaling expression data')\n",
    "sc.pp.scale(adata, max_value=10)\n",
    "\n",
    "# Combat adjustment                                                                                                                                                                                                           \n",
    "logging.info('Running combat')\n",
    "mod = patsy.dmatrix(\"~ disease_cov\", adata.obs, return_type=\"dataframe\")\n",
    "mod = mod.reset_index(drop=True)\n",
    "batch = adata.obs['batch_cov']\n",
    "batch = batch.reset_index(drop=True)\n",
    "adata.X = combat.combat(adata.X.transpose(), batch=batch, model=mod).transpose();\n",
    "\n",
    "logging.info(str('Data structure details: ' + str(adata)))\n",
    "logging.info('Removing samples that were processed twice...')\n",
    "# Remove samples that were processed twice and get rid of their 1.10 run\n",
    "remove = ['902289200_902289200', '1262_1262', '1270_1270', '1279_1279']\n",
    "# Remove samples that were processed twice and get rid of their 8.9 run\n",
    "# since the 8.9 run has the fewest number of cells, this adjustment is important for testing for cell proportion differences\n",
    "# going forward, we will want to combine batches for the same individuals\n",
    "remove2 = ['1472_1472', '1479_1479', '1480_1480', '1492_1492', '1522_1522', '1535_1535', '1602_1602', '1615_1615', '1621_1621', '1716_1716', '1726_1726', '1730_1730']\n",
    "keep_indices = ~((adata.obs.ind_cov.isin(remove)) & (adata.obs.batch_cov == \"lupus1.10\"))\n",
    "keep2_indices = ~((adata.obs.ind_cov.isin(remove2)) & (adata.obs.batch_cov == \"lupus8.9\"))\n",
    "adata = adata[(keep_indices & keep2_indices)]\n",
    "logging.info(str('Data structure details: ' + str(adata)))\n",
    "# Unique list of individuals\n",
    "people = np.unique(adata.obs['ind_cov'].values.tolist())\n",
    "# Allocate space for total PMBCs per individual.\n",
    "total_pbmcs = dict.fromkeys(people)\n",
    "for p in people:\n",
    "    total_pbmcs[p] = len(adata.obs_names[adata.obs['ind_cov'] == p])\n",
    "adata.uns['total_pbmcs'] = total_pbmcs\n",
    "logging.info('Saving processed data')\n",
    "# Remove .var information we never use.\n",
    "for key in list(adata.var.keys()):\n",
    "    del adata.var[key]\n",
    "logging.info(str('Structure details: ' + str(adata)))\n",
    "\n",
    "# Switch name nomenclature to be consistent.\n",
    "adata.obs['ind_cov'] = adata.obs['ind_cov'].astype('object')\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='1586_1586'] = 'FLARE002'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='904194200_904194200'] = 'FLARE003'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='1763_1763'] = 'FLARE008'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='1754_1754'] = '1754'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='1771_1771'] = '1771'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='1791_1791'] = '1791'\n",
    "adata.obs.ind_cov[adata.obs.ind_cov=='1804_1804'] = '1804'\n",
    "\n",
    "# Swith disease covariate convention from healthy, TR, UNT, to Healty, Managed, TR, UNT\n",
    "# Cross-sectional individuals\n",
    "CrossX = ['1240', '1754', '1771', '1791', '1804']\n",
    "adata.obs['disease_cov'] = adata.obs['disease_cov'].astype('object')\n",
    "adata.obs['disease_cov'][adata.obs['ind_cov'].isin(CrossX)] = 'Managed'\n",
    "adata.obs['disease_cov'][adata.obs['disease_cov']=='HEALTHY'] = 'Healthy'\n",
    "adata.obs['disease_cov'][adata.obs['disease_cov']=='TR'] = 'Treated'\n",
    "adata.obs['disease_cov'][adata.obs['disease_cov']=='UNT'] = 'Flare'\n",
    "\n",
    "# Add follow up therapy covariates to adata\n",
    "# No therapy data for 1791, FLARE018\n",
    "# Indicate steroid therapy.\n",
    "steroid = np.zeros(len(adata.obs_names))\n",
    "steroidpt = ['FLARE008', '1771','1804', 'FLARE003', 'FLARE004', 'FLARE006',\n",
    " 'FLARE007', 'FLARE009', 'FLARE011', 'FLARE013', 'FLARE015', 'FLARE017', 'FLARE020']\n",
    "steroid[adata.obs['ind_cov'].isin(steroidpt)] = 1\n",
    "# Indicate plaquenil therapy.\n",
    "plaquenil = np.zeros(len(adata.obs_names))\n",
    "plaquenilpt = ['1240', '1754', 'FLARE008', '1771', 'FLARE003','FLARE005', 'FLARE014', \n",
    "               'FLARE009', 'FLARE011', 'FLARE017', 'FLARE020']\n",
    "plaquenil[adata.obs['ind_cov'].isin(plaquenilpt)] = 1\n",
    "# Indicate cellcept therapy.\n",
    "cellcept = np.zeros(len(adata.obs_names))\n",
    "cellceptpt = ['1240', '1771', 'FLARE004', 'FLARE005', 'FLARE020']\n",
    "cellcept[adata.obs['ind_cov'].isin(cellceptpt)] = 1\n",
    "# Indicate rituxumab therapy.\n",
    "rituxumab = np.zeros(len(adata.obs_names))\n",
    "rituxumabpt = ['FLARE013', 'FLARE003', 'FLARE002']\n",
    "rituxumab[adata.obs['ind_cov'].isin(rituxumabpt)] = 1\n",
    "# Indicate cyclophosphamide therapy.\n",
    "cyclophosphamide = np.zeros(len(adata.obs_names))\n",
    "cyclophosphamidept = ['FLARE011', 'FLARE013', 'FLARE015', 'FLARE019']\n",
    "cyclophosphamide[adata.obs['ind_cov'].isin(cyclophosphamidept)] = 1\n",
    "adata.obs['steroid'] = steroid\n",
    "adata.obs['plaquenil'] = plaquenil\n",
    "adata.obs['cellcept'] = cellcept\n",
    "adata.obs['cyclophosphamide'] = cyclophosphamide\n",
    "adata.obs['rituxumab'] = rituxumab\n",
    "# Save both to active analysis file and also no_norm file for publication.\n",
    "adata.write(saveno_norm_path, compression=\"gzip\")\n",
    "adata.write(savepath, compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run basic analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Estimating louvain cluster identities for gene expression values.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing PCA with n_comps = 50\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:PCA complete.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    finished (0:00:24.25)\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 50\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:KNN complete.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    finished (0:01:01.00)\n",
      "computing Diffusion Maps using n_comps=15(=n_dcs)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:diffmap complete.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    eigenvalues of transition matrix\n",
      "    [1.         0.9991194  0.99885595 0.9972867  0.99322224 0.9925152\n",
      "     0.9915735  0.98981833 0.98870987 0.98436844 0.979085   0.97838694\n",
      "     0.97352386 0.96651393 0.965863  ]\n",
      "    finished (0:00:12.38)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:48.61)\n",
      "running Leiden clustering\n",
      "    finished (0:03:49.61)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Louvain complete.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing UMAP\n",
      "    using 'X_pca' with n_pcs = 50\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:UMAP complete.\n",
      "INFO:root:Making .obs into categories\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    finished (0:03:58.60)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Basic analysis complete.\n"
     ]
    }
   ],
   "source": [
    "#######################\n",
    "# Louvain and friends #\n",
    "#######################\n",
    "# Set parameters\n",
    "intialization = 1\n",
    "n_components = 20\n",
    "resolution = 3\n",
    "# Run louvain clustering on theoretical future gene expression per cell\n",
    "logging.info('Estimating louvain cluster identities for gene expression values.')\n",
    "sc.pp.pca(adata, random_state=intialization, svd_solver='arpack')\n",
    "logging.info('PCA complete.')\n",
    "sc.pp.neighbors(adata, random_state=intialization)\n",
    "logging.info('KNN complete.')\n",
    "sc.tl.diffmap(adata)\n",
    "logging.info('diffmap complete.')\n",
    "sc.tl.louvain(adata, random_state=15, resolution=resolution)\n",
    "sc.tl.leiden(adata, random_state=intialization, resolution=resolution)\n",
    "logging.info('Louvain complete.')\n",
    "sc.tl.umap(adata, random_state=intialization)\n",
    "logging.info('UMAP complete.')\n",
    "logging.info('Making .obs into categories')\n",
    "adata.strings_to_categoricals()\n",
    "adata.write(savepath, compression=\"gzip\")\n",
    "logging.info('Basic analysis complete.')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
